tests/core/daemon/test_daemon.py

```diff
@@ -72,55 +72,56 @@ async def num_connections():
-                    await queue.put(message)
-                elif msg.type == aiohttp.WSMsgType.PING:
-                    await ws.pong()
-                elif msg.type == aiohttp.WSMsgType.PONG:
-                    continue
-                else:
-                    if msg.type == aiohttp.WSMsgType.CLOSE:
-                        await ws.close()
-                    elif msg.type == aiohttp.WSMsgType.ERROR:
-                        await ws.close()
-                    elif msg.type == aiohttp.WSMsgType.CLOSED:
-                        pass
-
-                    break
-
-        read_handler = asyncio.create_task(reader(ws, message_queue))
-        data = {}
-        payload = create_payload("get_blockchain_state", data, service_name, "chia_full_node")
-        await ws.send_str(payload)
-
-        await asyncio.sleep(5)
-        blockchain_state_found = False
-        while not message_queue.empty():
-            message = await message_queue.get()
-            if message["command"] == "get_blockchain_state":
-                blockchain_state_found = True
-
-        await ws.close()
-        read_handler.cancel()
+        ) as ws:
+            service_name = "test_service_name"
+            data = {"service": service_name}
+            payload = create_payload("register_service", data, service_name, "daemon")
+            await ws.send_str(payload)
+            message_queue = asyncio.Queue()
+
+            async def reader(ws, queue):
+                while True:
+                    msg = await ws.receive()
+                    if msg.type == aiohttp.WSMsgType.TEXT:
+                        message = msg.data.strip()
+                        message = json.loads(message)
+                        await queue.put(message)
+                    elif msg.type == aiohttp.WSMsgType.PING:
+                        await ws.pong()
+                    elif msg.type == aiohttp.WSMsgType.PONG:
+                        continue
+                    else:
+                        if msg.type == aiohttp.WSMsgType.CLOSE:
+                            await ws.close()
+                        elif msg.type == aiohttp.WSMsgType.ERROR:
+                            await ws.close()
+                        elif msg.type == aiohttp.WSMsgType.CLOSED:
+                            pass
+
+                        break
+
+            read_handler = asyncio.create_task(reader(ws, message_queue))
+            try:
+                data = {}
+                payload = create_payload("get_blockchain_state", data, service_name, "chia_full_node")
+                await ws.send_str(payload)
+
+                await asyncio.sleep(5)
+                blockchain_state_found = False
+                while not message_queue.empty():
+                    message = await message_queue.get()
+                    if message["command"] == "get_blockchain_state":
+                        blockchain_state_found = True
+            finally:
+                read_handler.cancel()
```

Comment:
Now we get into a funny area.  You should await the task as well.  But, you want to ignore the `asyncio.CancelledError` that is going to come out of it.  You fed it in, you know you want to cancel up to this point.  But...  you will be awaiting so this task could be getting cancelled as well.  How do you tell which task was cancelled?  One cancellation you want to ignore, the other you want to propagate.  https://docs.python.org/3.10/library/asyncio-task.html#asyncio.Task.cancel in Python 3.9 they added a `msg=` parameter so we could make distinguishable cancellation requests.  I really need to look into whether we can ease into use of AnyIO piece by piece.

Reply:
so.. `await` on a task throws `CancelledError` when it's cancelled? that seems pretty unintuitive. The handle to the task isn't a future, right? or does the return value of the task get passed back though the handle? if so, I suppose it makes sense to receive the exception there too.

One would think there's a way to wait on a future, separate from asking for its value though.

Reply:
Yeah, awaiting a task is one way you 'reattach' it and accept it's output, either return or exception.  `x = await af()` is at least pretty similar to `x = await asyncio.create_task(af())` in terms of running the coroutine returned by `af()` and feeding out the return value or exceptions.  You can also `asyncio.gather()`, `asyncio.wait()`, `asyncio.wait_for()`, and probably a few others depending on various needs.

https://docs.python.org/3.10/library/asyncio-task.html#task-object
> A [Future-like](https://docs.python.org/3.10/library/asyncio-future.html#asyncio.Future) object that runs a Python [coroutine](https://docs.python.org/3.10/library/asyncio-task.html#coroutine).